src/proto/loaded_image.rs

```diff
@@ -53,6 +53,19 @@ impl LoadedImage {
         self.device_handle
     }
 
+    /// Return a NULL-terminated Path string including directory and file names.
+    pub fn file_path(&self) -> &CStr16 {
```

Comment:
I think it might make sense to have this return a [`&DevicePath`](https://docs.rs/uefi/latest/uefi/proto/device_path/struct.DevicePath.html) instead of `&CStr16`, since I don't think there's a guarantee in the spec about which specific type of device path is going to be returned here.

I'm also not sure if the `file_path` field is guaranteed to be non-null (for example, what if the image was loaded from a buffer instead of a path, and the path wasn't set?). If I have that right, the return type can be `Option<&DevicePath>`.

Reply:
> I'm also not sure if the file_path field is guaranteed to be non-null (for example, what if the image was loaded from a buffer instead of a path, and the path wasn't set?). If I have that right, the return type can be Option<&DevicePath>.

I do some test for it now.

(the code inside the println block is similar to the PR commit.)

(the method to get file path using the 5th field of `LoadedImage`)

First step, create a child image which will print the it's `file path`:

![t1](https://user-images.githubusercontent.com/84341772/161360365-2e6845ef-1831-4612-9c71-89ea3a9e4392.png)

Then I compile it and then place the PE file into `src`.
Now let's change some code:

- include the bytes of previous compiled.
- Then print itself's file path (as a parent and loaded by firmware directly)
- load image from the bytes(buffer) without `file_path`.
- then start it

![t2](https://user-images.githubusercontent.com/84341772/161360719-ee4b532e-8b43-4ca1-bf41-ce9ccba91b8e.png)

Next, compile what we changed.

The last step is do some test, the script for testing are as follow:

``` shell
#!/bin/sh

dd if=/dev/zero of=test.img bs=512 count=532480
mkfs.fat -F32 test.img
mkdir -p mnt
mount test.img mnt
mkdir -p mnt/efi/boot
cp target/x86_64-unknown-uefi/release/boot.efi mnt/efi/boot/bootx64.efi
umount mnt
qemu-system-x86_64 -bios /usr/share/ovmf/OVMF.fd -drive file=test.img,format=raw
```

Then the result:

![t4](https://user-images.githubusercontent.com/84341772/161360863-b9939fd0-4784-43a7-ac81-4fcbc7994979.png)

# Summary
- the parent that directly loaded by UEFI has print its correct file path.
- load the image from buffer without point out the `file_path` has print a empty path.

> I think it might make sense to have this return a [&DevicePath](https://docs.rs/uefi/latest/uefi/proto/device_path/struct.DevicePath.html) instead of &CStr16, since I don't think there's a guarantee in the spec about which specific type of device path is going to be returned here.

# Still updating...

Farther seps:
- print the `DevicePathHeader` of file_path and see it's type (should be 0x04 `MEDIA` and sub type should be 
0x04 `FilePath`) & the length of data following the header (test if it's length equal 5 (file_path.length - sizeof(file_path) == 1)).

Reply:
Thanks for testing that out! I think though that we have to rely on what is actually guaranteed in the spec, because even if one UEFI implementation (such as EDK2) always sets the file path, and always uses a device path type of File Path Media Device Path, we can't assume other implementations will.

Here's what I see as the relevant bits in the spec:
* 9.1 EFI Loaded Image Protocol defines `FilePath` as "A pointer to the file path portion specific to DeviceHandle that the EFI Image was loaded from." I read that as saying that the `FileHandle` points to a device path node with the device path passes to `EFI_BOOT_SERVICES.LoadImage`.
* 7.4 EFI_BOOT_SERVICES.LoadImage goes into some more details about which protocols get used to load the image data depending on the device path, but I don't see any specific guarantee that the device path must be a  File Path Media Device Path.

So I think the right thing to do here is to return `Option<&DevicePath>`, and then leave it up to the caller to check what type the resulting `DevicePath` is.

Admittedly the current API of `DevicePath` doesn't make this quite as convenient as we might want, I've just filed a couple new issues to address that:
* https://github.com/rust-osdev/uefi-rs/issues/400
* https://github.com/rust-osdev/uefi-rs/issues/401

Reply:
> So I think the right thing to do here is to return Option<&DevicePath>, and then leave it up to the caller to check what type the resulting DevicePath is.

It's done now, and I also add a method to `checked_get` the path string (src/table/boot.rs).😃